### ISSUE TITLE
[Feature]: MacOS as server support

### ISSUE THREAD

**Issue by Tahul** [coollabsio/coolify#2773]
### Description

I have some servers running on MacOS ; they also already run Docker perfectly fine.

Also, it is fairly easy to give a user root access on MacOS.

I feel there is no major blocker for that task, but I would love to know more about it if there is some.

I would happily add 50$ of bounty if someone confirms there is no major blocker for that task.

### Minimal Reproduction (if possible, example repository)

I can provide access to a MacOS server if needed; currently I get this screen when trying to validate a server:

![CleanShot 2024-07-08 at 23 57 46@2x](https://github.com/coollabsio/coolify/assets/1385263/2c48eb9e-45ff-4525-a4cb-cce2345fc9b1)

Note I can properly add and SSH login to the server, but I cannot setup the Coolify env.

### Version

cloud (beta.306)

**Comment by Xiloe** [comment]
/etc/os-release doesn't exist on macos hence the error (no such file or directory)

looks like it's called both in : https://github.com/coollabsio/coolify/blob/e9158b7305bb9253fe75a0e76ef34af62f75199b/app/Jobs/ServerStatusJob.php#L73

and : https://github.com/coollabsio/coolify/blob/e9158b7305bb9253fe75a0e76ef34af62f75199b/app/Models/Server.php#L873

A mac os alternative would be :
`$ sysctl kern.version`
`$ sysctl kern.ostype`
`$ sysctl kern.osrelease`
`$ sysctl kern.osrevision`

or `$ sw_vers`

though only changing those wouldn't work, a lot of stuff relies on this file, e.g. the installer script (https://github.com/coollabsio/coolify/blob/e9158b7305bb9253fe75a0e76ef34af62f75199b/scripts/install.sh#L13)

and it's not even referenced in the supported os list : https://github.com/coollabsio/coolify/blob/e9158b7305bb9253fe75a0e76ef34af62f75199b/bootstrap/helpers/constants.php#L39

that said, I'm not sure this is the only thing you'd need to fix for proper macos support, a bunch of other stuff might break

**Comment by Tahul** [comment]
From what I've understood of Coolify "server" setup ; it's mainly a bunch of Docker containers, so I guess that would be achievable as Docker runs perfectly fine on MacOS?

**Comment by Xiloe** [comment]
Yeah, I never used the cloud version nor tried to deploy coolify on my mac, so I can't really tell, but from the error message and the references in the source that's all I can tell. hope will come handy if you take a try at it :)

**Comment by Xiloe** [comment]
I actually went on my macbook to test this out, first of all I was able to run the docker compose and have coolify run here, as I don't have a macos server I can't try the Validate & Install thing, but looking at the code, here is the general flow :

When you click on the button, it calls the `validateServer` function: https://github.com/coollabsio/coolify/blob/e9158b7305bb9253fe75a0e76ef34af62f75199b/resources/views/livewire/server/form.blade.php#L39

which is basically dispatching an `init` message: https://github.com/coollabsio/coolify/blob/e9158b7305bb9253fe75a0e76ef34af62f75199b/app/Livewire/Server/Form.php#L166

the actual `init functions seems to be here : https://github.com/coollabsio/coolify/blob/e9158b7305bb9253fe75a0e76ef34af62f75199b/app/Livewire/Server/ValidateAndInstall.php#L46-L59

which at the end calls `validateConnection`, which itself calls `validateOS` which itself calls the `$ cat /etc/os-release` to check the os ID

https://github.com/coollabsio/coolify/blob/e9158b7305bb9253fe75a0e76ef34af62f75199b/app/Livewire/Server/ValidateAndInstall.php#L94

https://github.com/coollabsio/coolify/blob/e9158b7305bb9253fe75a0e76ef34af62f75199b/app/Livewire/Server/ValidateAndInstall.php#L97-L106

the 500 error comes from the fact that they most likely do not check if the command actually runs successfully, since the error is "no such file or directory", instead of the expected : "Server OS type is not supported. Please install Docker manually before continuing: <a target="_blank" class="underline" href="https://docs.docker.com/engine/install/#server">documentation</a>."

**Comment by Xiloe** [comment]
I tried to implement a solution for the docker installation issue, which is the main issue as coolify runs fine on a macbook as is. The issue now is finding a good alternative for `systemctl` based commands to install/enable docker on the server(macos machine). Since there is no default package manager on macos, brew would probably be the best option imo, but then it means adding an install step on macos to install brew, then docker, to be able to use `brew services`.

I also think using `uname -s` to detect if it's a macos machine (Darwin and not Linux/...) should be consistent and good enough since Darwin is an apple only thing.

**Comment by Tahul** [comment]
I would love `brew services` to be the alternative to systemd as this is what I use myself for my existing workaround.

Also I think it is the most commonly used alternative on MacOS.

**Comment by Tahul** [comment]
Really nice explanations about implications of such task; thank you so much for being so descriptive.

**Comment by Xiloe** [comment]
I might be wrong on this one, but I don't think there is a way to get the docker engine/daemon with brew unless you install Docker Desktop and run it, you may know more than me on this @Tahul , because it would annoy me if I had to use Docker Desktop instead of a brew service. I always used the full desktop install as I never got it to work without, I know there are alternatives like colima/podman and whatnot but I'd like to use docker if possible to avoid using other third parties that might cause issues later on.

**Comment by Xiloe** [comment]
yeah even the install script that they use from rancher is just redirecting to docker desktop
<img width="884" alt="image" src="https://github.com/coollabsio/coolify/assets/51050533/810faa6b-6ced-42e1-a1f3-e990922fd96a">
that's a bummer

**Comment by Xiloe** [comment]
> I do use Colima on MacOS to run Docker w/o Docker Desktop:
> 
> https://github.com/abiosoft/colima
> 
> It might be a great solution for such kind of "docker-wrapper" like Coolify Server?

it was my second option anyways so I guess I'll go with that, I don't want to get a full install in there, plus it's going to be annoying to restart etc, as it's the desktop app, I'd basically have to kill the process and relaunch it which isn't ideal

**Comment by Xiloe** [comment]
The _last_ issue I have right now is begin able to download the correct docker engine version that coolify wants, it's a bit of a pain but after this is figured out I guess I'll have a first PR ready to be tested on an actual server, I don't have any so I'll let you do that and hopefully it all goes well :)

**Comment by jSchenkel** [comment]
Hello! I'm currently developing an AI tool that investigates issues, and I used this issue as a test to see if it could help. Below are the findings, which I hope might assist in resolving this issue faster.

**Boosterbot.ai Findings:**

The investigation identified key areas that need to be updated to support MacOS servers within the Coolify application. The `Server.php` model class should be updated to recognize MacOS as a valid operating system. The SSH connection handling code in `remoteProcess.php` and the `ExecuteRemoteCommand` trait should be reviewed to ensure compatibility with MacOS. The environment setup scripts (`install.sh` and `upgrade.sh`) and configuration management classes (`ConfigureCloudflared` and `UpdateCoolify`) should be updated to include support for MacOS-specific configurations. Additionally, the `coolify.php` configuration file should be reviewed to ensure that any environment-specific settings include support for MacOS.
1. **Review the code responsible for server validation to understand how servers are recognized and validated within the Coolify application.**
The `Server.php` file in the `app/Models` directory contains methods like `validateOS`, `validateConnection`, and `validateDockerEngine` that are crucial for server validation. The `validateOS` method should be updated to recognize MacOS as a valid operating system. Additionally, the `config/coolify.php` file should be checked to ensure that any OS-specific configurations include support for MacOS.
2. **Investigate the SSH connection handling code to ensure that SSH connections to MacOS servers are properly managed.**
The `remoteProcess.php` file in the `bootstrap/helpers` directory contains functions like `generateSshCommand`, `instant_remote_process`, and `refresh_server_connection` that manage SSH connections. These functions should be reviewed to ensure compatibility with MacOS. Additionally, the `ExecuteRemoteCommand` trait in the `app/Traits` directory should be checked to ensure that remote command execution works seamlessly on MacOS servers.
3. **Examine the environment setup process to identify any OS-specific configurations or checks that might be causing issues with MacOS servers.**
The `install.sh` and `upgrade.sh` scripts in the `scripts` directory handle the environment setup and upgrade process. These scripts should be reviewed to ensure they include support for MacOS. Additionally, the `ConfigureCloudflared` and `UpdateCoolify` classes in the `app/Actions/Server` directory should be checked to ensure that they handle MacOS-specific configurations correctly. The `coolify.php` configuration file should also be reviewed to ensure that any environment-specific settings include support for MacOS.

These suggestions are automatically generated based on the context of the issue description and repository code. If you have any feedback or want to try it on your own project, my email is on my profile. Thanks!

**Comment by Tahul** [comment]
@jSchenkel could you remove your comment? I feel it is VERY disrespectful for the work of @Xiloe to come up with unsolicited AI generated content while some real people are pursuing the work in a serious manner.

**Comment by jSchenkel** [comment]
@Tahul My apologies, my intention was to be helpful and share my tool. Given that this is a bounty and @Xiloe has already done great work on the issue, I realize now that I made a mistake sharing the AI results. Deleted. Sorry @Xiloe !

**Comment by Tahul** [comment]
@jSchenkel ; no worries, the results seemed great tbh! Happy we come to the same conclusion, good job with your project!

**Comment by Xiloe** [comment]
Hi, I wasn't able to work on anything yesterday, will try to make a commit today :)

**Comment by Xiloe** [comment]
@Tahul Would you mind trying [this branch](https://github.com/Xiloe/coolify/tree/macos-server-support) on your macos server by any chance?

It's not perfect at all, especially the docker install script that I could test, it doesn't even use the version specified by `$dockerVersion` yet, though the rest should be working fine, maybe try to have docker installed already if it fails miserably. Hope you can give my some feedback :)

**Comment by Tahul** [comment]
@Xiloe thank you so much ; is it possible to use your server version while on a Cloud setup?

**Comment by Tahul** [comment]
Also now that there has been great efforts achieved on this task; would you mind taking a look/giving a feedback on the purpose of this issue @andrasbacsai? 🙏

**Comment by Xiloe** [comment]
> @Xiloe thank you so much ; is it possible to use your server version while on a Cloud setup?

Not sure that's possible, you'd have to selfhost my branch and then add your macos server from there

**Comment by Tahul** [comment]
Alright ; I have a sandbox server I'll try it from there.

Not sure I'd have the bandwith for it today but maybe in the beginning of next week.

**Comment by peaklabs-dev** [comment]
@Tahul and @Xiloe I see there has been some great progress and even a WIP implementation here: https://github.com/coollabsio/coolify/compare/main...Xiloe:coolify:macos-server-support. My main issue with adding mac os support is still the docker issue, as MacOs does not have docker APIs, there would be the need for docker desktop (too slow), or Colima, or maybe Orbstack for better performance, but all of these are essentially running a Linux VM with Docker? so why use MacOS as a server at all, when you then are running linux anyways? 

If there is still enough interest, I may be able to bring this up/ work on a full implementation in core Coolify.

**Comment by Xiloe** [comment]
Hey @peaklabs-dev I had the same thought pretty much, docker on macos is kind of a mess if you're not using the desktop app, and all the other solutions like colima etc runs a VM which is not ideal either imo, at this point using linux is just easier.

**Comment by CavinHuang** [comment]
wait

**Comment by peaklabs-dev** [comment]
@CavinHuang What did you want to say?

@Tahul Since you seem to be one of the main Mac guys, I would really like to hear the reason why this would be necessary (see my opinion in the comment): https://github.com/coollabsio/coolify/issues/2773#issuecomment-2311245860)

**Comment by peaklabs-dev** [comment]
After thinking about it, I think it makes sense to support MacOs since we will also support Windows/Windows Server. MacOs is not the most recommended way, but at least it should work.
I am thinking about using Orbstack or maybe https://asahilinux.org/ to add Mac OS support (after v4 stable releases).

**Comment by arttus** [comment]
It would be cool to support docker desktop so users on Mac can try the stack on our local dev machines. Orbstack would be fine as well, but I'm guessing more devs already have Docker desktop running.

**Comment by iRonin** [comment]
You can do `sudo touch /etc/os-release` but what then with the Docker detection?

**Comment by gleatd01** [comment]
Sounds like orbstack might enable this

**Comment by saheljalal** [comment]
I have a few mac minis which I've been manually running services (Docker, scripts, etc.) on over the years and discovered this project recently. It seemed like it would make everything so much easier by being able to deploy things my own machines and track from a Vercel like self-hosted platform.

Please do consider adding support for macOS as a server, as it would be a fantastic way to make this project significantly more useful out of the box.

**Comment by gleatd01** [comment]
> I have a few mac minis which I've been manually running services (Docker, scripts, etc.) on over the years and discovered this project recently. It seemed like it would make everything so much easier by being able to deploy things my own machines and track from a Vercel like self-hosted platform.
> 
> Please do consider adding support for macOS as a server, as it would be a fantastic way to make this project significantly more useful out of the box.

I am getting another machine next week and will be repurposing an Apple M2 to go after the bounty for this one. Hopefully, I'll be able to make it work.

**Comment by sagararyal** [comment]
Happy to offer remote access to a Virtual MacOS M4 (hosted in DE) if anyone needs it for development purpose.
email: sagar [at] aryal [dot] me

**Comment by rossanmol** [comment]
Did anyone manage to setup coolify on mac?

**Comment by gleatd01** [comment]
> Did anyone manage to setup coolify on mac?

I haven't had time yet, but I may put the effort in this weekend.

**Comment by MimouniSaad** [comment]
@

> > Did anyone manage to setup coolify on mac?
> 
> I haven't had time yet, but I may put the effort in this weekend.

Can you please update us ?

**Comment by liamyehudai** [comment]
> > Did anyone manage to setup coolify on mac?
> 
> I haven't had time yet, but I may put the effort in this weekend.

I would also be interested in an update, this would be incredibly useful for my lab

**Comment by shiv19** [comment]
I recently setup coolify on my mac.
I installed it inside a headless VM running with Colima.
I used the default docker VM that you can spin up with colima. (Make sure you forward your host ssh agent to colima)
Then I did `colima ssh` to ssh into it.
Installed coolify in there.
Setup cloudflare tunnel service in coolify and I was able to add wild card domain support for services running in coolify.
Everything works smoothly.
I've allocated 60gb storage and 2gb ram to the colima vm and it is easily hosting 8 hobby projects with no hiccups.

@Tahul until there is official macos support, I think this is the best way to set it up.

**Comment by cefege** [comment]
Looking forward to Mac suppot

**Comment by sagararyal** [comment]
I think an end is coming to this Issue: https://github.com/coollabsio/coolify/issues/5685#issuecomment-2828959198

> * We will probably drop `MacOS` and `Windows` support altogether, as it is just a huge burden to maintain. Of course, you can still use your Mac hardware if you install Linux or use a VM or dual-boot Linux on your Mac.
> * Reasons:
>   
>   * The main reason is that MacOS and Windows lack native support for Docker/Podman. To get Docker support, you need Docker Desktop or Orbstack, which just run a Linux VM in the background, which is a small performance overhead but then you could also just use Linux.
>   * Also there is a hole impossible set of issues on MacOS with bind mount and volumes because of MacOS filesystem protection (and trust me I tried to fix it for my dev environment but no luck).
>   * And for Windows it has a hole other set of issues with the Proxy and very poor performance WSL.

**Comment by Q00** [comment]
/attempt #2773

Added macOS support for Coolify installer script

The original install.sh script doesn't work on macOS because it relies on /etc/os-release which doesn't exist on macOS. This PR adds a separate install-macos.sh script

**Comment by stewnight** [comment]
Might be something to revist with Apple's new container tool

https://github.com/apple/containerization

**Comment by gleatd01** [comment]
> I recently setup coolify on my mac. I installed it inside a headless VM running with Colima. I used the default docker VM that you can spin up with colima. (Make sure you forward your host ssh agent to colima) Then I did `colima ssh` to ssh into it. Installed coolify in there. Setup cloudflare tunnel service in coolify and I was able to add wild card domain support for services running in coolify. Everything works smoothly. I've allocated 60gb storage and 2gb ram to the colima vm and it is easily hosting 8 hobby projects with no hiccups.
> 
> [@Tahul](https://github.com/Tahul) until there is official macos support, I think this is the best way to set it up.

I think Colima would be worth a try because of the docker support. Might make this easier and worth revisiting.

**Comment by launderex619** [comment]
> I recently setup coolify on my mac. I installed it inside a headless VM running with Colima. I used the default docker VM that you can spin up with colima. (Make sure you forward your host ssh agent to colima) Then I did `colima ssh` to ssh into it. Installed coolify in there. Setup cloudflare tunnel service in coolify and I was able to add wild card domain support for services running in coolify. Everything works smoothly. I've allocated 60gb storage and 2gb ram to the colima vm and it is easily hosting 8 hobby projects with no hiccups.
> 
> [@Tahul](https://github.com/Tahul) until there is official macos support, I think this is the best way to set it up.

I did this and works perfectly!  Some things to add:

- To attach the ssh agent is `colima start --ssh-agent`
- The exposed ssh port is not 22 (I struggled in this step due to ssh required connections to launch services), when installing coolify it pops out a ssh port in the instalation steps that (in my case) disappeared kind of fast, so I reinstall and try with that port and worked perfectly (60247 in my case).

**Comment by Cinzya** [comment]
If anyone wants to contribute to the Coolify Docs and add a few lines about this, we are happy to merge and include this to the Docs :)

**Comment by andrasbacsai** [comment]
I am closing this as mac support is not in scope atm, but PR's are welcome.

**Comment by mrtnsch** [comment]
@Tahul  @shiv19 thanks for your comments, this pointed me in the right direction for setting everything up on my mac mini. I have opened a PR documenting my approach using colima in detail: https://github.com/coollabsio/coolify-docs/pull/565
